### PR TITLE
Use sudo instead

### LINKED ISSUE #85: Haskell available only for root
started this issue for the @DaneWeber  PR #74

### PR DESCRIPTION
**Pull request by jcbhmr**
Closes #85 with @jcbhmr's changes.

### CODE REVIEW

**Comment by jcbhmr** [comment]
_This is a discussion unrelated to this PR_

What do you think about piping to sh like `curl ... | sh` vs piping to `/bin/sh` vs [downloading it and using the hash-bang line](https://stackoverflow.com/questions/31717305/run-a-remote-script-using-shebang-in-linux)?

```sh
# Basically do whatever the installer instructions say
curl ... | sh
curl ... | bash
curl ... | python3
```

Use the hash-bang by downloading the file and `chmod +x`-ing it
```sh
curl -o INSTALLFILE https://...
chmod +x INSTALLFILE
./INSTALLFILE
```

**Comment by danielbraun89** [comment]
probably following  the instructions given by whatever website is safest,  as it probably knew which shell is required for the specific syntax

If it had a shebang in it, I guess it would be respected and replaced to it regardless of the shell you launched it with, so also good

**Comment by DaneWeber** [comment]
I'm a little torn because I see some trade-offs. Piping to a shell (`... | sh`) is presumably in the official install instructions primarily to make installation easier: just copy-paste one line instead of following multiple instructions. Since we're writing scripts/programs here, however, fitting everything on a single line is generally harder to read and understand.

That said, my sense is that the priority in this repo is to optimize for simplicity and ease of maintenance since we want contributors and will likely only discover that a given Feature needs an update when an upstream dependency breaks the build. As such, I really ❤️ the link to the source of the install instructions and following it verbatim. If this Feature fails to install in the future, it will be very easy to check whether the GHCup folks have changed the install method.

**Comment by DaneWeber** [comment]
Hey @jcbhmr, I'm having a terrible time right now with getting different versions to install via the settings specified in `.devcontainer.json`. I'm 99% sure that the various env vars set in the script are _not_ making it into the `sudo` command. I've been trying to pass them in, and thought I had done so (confirming with echoing them into a text file), but I can't get the GHCup installer to respect them.

My question to you: did you verify that if you set a value of `9.2.4` for `ghcVersion` etc., that version is what is actually installed? I may just have something messed up with my local situation.

**Comment by jcbhmr** [comment]
> Hey @jcbhmr, I'm having a terrible time right now with getting different versions to install via the settings specified in `.devcontainer.json`. I'm 99% sure that the various env vars set in the script are _not_ making it into the `sudo` command. I've been trying to pass them in, and thought I had done so (confirming with echoing them into a text file), but I can't get the GHCup installer to respect them.
> 
> My question to you: did you verify that if you set a value of `9.2.4` for `ghcVersion` etc., that version is what is actually installed? I may just have something messed up with my local situation.

Ah! No I did not verify that the `$GHC_VERSION` was passed through. It's not 😞. **BUT** I have good news! It's only because I didn't know that it needed to. In [another instance very similar](https://github.com/devcontainers-contrib/features/issues/80#issuecomment-1327910397), I used some env vars [and commented why some are the way they are](https://github.com/devcontainers-contrib/features/pull/86/files#diff-129f15384236356bbb9f97fbfe5f6986dc3b619fd4162b27125c478e0259a729R30).

```sh
# We run as the non-root user so to fix https://github.com/devcontainers-contrib/features/issues/80
# Note that we substitute SOME variables before evaluation, and some are
# substituted inside the $_REMOTE_USER shell. Particularily $HOME which needs to
# be from the $_REMOTE_USER, and $VERSION which needs to come from this script.
```

https://github.com/devcontainers-contrib/features/blob/main/src/pulumi/install.sh#L30

Basically, it's because ENV vars _don't pass through the `sudo` boundary_ since it's like a whole different user executing. Thus, if you want any to go through, (like I should have done 😢) you must:

```sh
sudo -iu "$_REMOTE_USER" <<EOF
  GHC_VERSION="$GHC_VERSION"
  CABAL_VERSION="$CABAL_VERSION"
  INCLUDE_STACK="$INCLUDE_STACK"
  ADJUST_BASHRC="$ADJUST_BASHRC"
  
  INSTALL_STACK_GHCUP_HOOK="${INSTALLSTACKGHCUPHOOK:-"true"}"
  # Note that I can use a \$ to avoid substituting NOW and leave that the other shell pass to do
  export STRING_WITH_DOLLAR_SIGN_IN_IT='\$GHC_VERSION'
EOF
```

I can open a PR & issue to fix this. I should have caught this earlier 😡. Oh well!

**Comment by DaneWeber** [comment]
@jcbhmr, I was trying to pass them in with the `--preserve-env=` flag on `sudo`, but your approach looks more reliable (since my approach was unwieldy and falling).

Note that you don't need the env vars at the top, but the ones set in https://github.com/devcontainers-contrib/features/blob/0495b96620759692a892b14b449f41acaeb93722/src/haskell/install.sh#L31-L45

One gotcha I can test is that the so-called boolean vars are true if set, so we'll need to make sure they don't get set to true simply by being included in the list.
